policy/aggregation_type.go

```diff
@@ -0,0 +1,274 @@
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package policy
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/m3db/m3metrics/generated/proto/schema"
+)
+
+// Supported aggregation types.
+const (
+	Unknown AggregationType = iota
+	Last
+	Lower
+	Upper
+	Mean
+	Median
+	Count
+	Sum
+	SumSq
+	Stdev
+	P50
+	P95
+	P99
+	P999
+	P9999
+	totalAggregationTypes = iota
+)
+
+const (
+	// AggregationIDLen is the length of the AggregationID.
+	// The AggregationIDLen will be 1 when totalAggregationTypes <= 64.
+	AggregationIDLen = (totalAggregationTypes-1)/64 + 1
+
+	aggregationTypesSeparator = ","
+)
+
+var (
+	// DefaultAggregationTypes is a default list of aggregation types.
+	DefaultAggregationTypes AggregationTypes
+
+	// DefaultAggregationID is a default AggregationID.
+	DefaultAggregationID AggregationID
+
+	// ValidAggregationTypes is the list of all the valid aggregation types
+	ValidAggregationTypes = []AggregationType{
+		Last,
+		Lower,
+		Upper,
+		Mean,
+		Median,
+		Count,
+		Sum,
+		SumSq,
+		Stdev,
+		P50,
+		P95,
+		P99,
+		P999,
+		P9999,
+	}
+
+	aggregationTypeStringMap map[string]AggregationType
+)
+
+func init() {
+	aggregationTypeStringMap = make(map[string]AggregationType, totalAggregationTypes)
+	for _, aggType := range ValidAggregationTypes {
+		aggregationTypeStringMap[aggType.String()] = aggType
+	}
+}
+
+// AggregationType defines a custom aggregation function.
+type AggregationType int
+
+// NewAggregationTypeFromSchema creates an aggregation type from a schema.
+func NewAggregationTypeFromSchema(input schema.AggregationType) (AggregationType, error) {
+	aggType := AggregationType(input)
+	if !aggType.IsValid() {
+		return Unknown, fmt.Errorf("invalid aggregation type from schema: %s", input)
+	}
+	return aggType, nil
+}
+
+// IsValid checks if an AggregationType is valid.
+func (a AggregationType) IsValid() bool {
+	return a > 0 && a < totalAggregationTypes
+}
+
+// IsValidForGauge if an AggregationType is valid for Gauge.
+func (a AggregationType) IsValidForGauge() bool {
+	switch a {
+	case Last, Lower, Upper, Mean, Count, Sum, SumSq, Stdev:
+		return true
+	default:
+		return false
+	}
+}
+
+// IsValidForCounter if an AggregationType is valid for Counter.
+func (a AggregationType) IsValidForCounter() bool {
+	switch a {
+	case Lower, Upper, Mean, Count, Sum, SumSq, Stdev:
+		return true
+	default:
+		return false
+	}
+}
+
+// IsValidForTimer if an AggregationType is valid for Timer.
+func (a AggregationType) IsValidForTimer() bool {
+	switch a {
+	case Last:
+		return false
+	default:
+		return true
+	}
+}
+
+// ParseAggregationType parses an aggregation type.
+func ParseAggregationType(str string) (AggregationType, bool) {
```

Comment:
I think the convention is to return an error instead of a bool for `Parse...` like functions.